mne_bids/mne_bids.py

```diff
@@ -537,8 +579,16 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
     config : str | None
         A path to the configuration file to use if the data is from a BTi
         system.
-    overwrite : bool
-        If the file already exists, whether to overwrite it.
+    write_mode : str, one of ('append', 'overwrite', 'error')
+        How any existing files should be handled when new data is produced.
+        Defaults to 'append'.
+        - 'append': Any file that already exists will be overridden with new
```

Comment:
I feel this is dangerous as a default. See also [similar comments](https://github.com/mne-tools/mne-python/pull/5589/files#r223330583) on MNE-Python. What do you think @teonbrooks and @sappelhoff ?

Reply:
I agree that overwriting should not be the default option. Making "error" the default is arguably safer.

Reply:
ok, will change that and rebase...

Reply:
actually, one issue I realise. If `write_mode = 'error'` by default then this will carry through to the writing of the scans and participants files. Is this the behaviour we want? An error would be raised if this file exists even if none of the other files exist (which would generally be considered the behaviour we want...)
I can't really think of a good way to handle this... Do we even really want the attempted writing of these files to raise an error at any point??

Reply:
hmm ... the other possibility is to raise an error in the `'append'` mode if the file already exists, make that the default and get rid of the `'error'` mode.

Reply:
But if we do that, we are back to two options -- which makes me think we only  need `overwrite=True/False`

Reply:
but we wouldn't want it to raise an error in the append mode if it already exists, we want it to append...
I think the problem is most thing have two different options, but not everything has the same two options...

Reply:
> but we wouldn't want it to raise an error in the append mode if it already exists, we want it to append...

but only for those two specific files right? Or do you mean you want to be able to incrementally add files to the directory? That is -- it will check if the file exists and add it only if it doesn't. Is that what the current behavior is right now?

Reply:
yep, only for those two files.
Most files have only one set of data in them, so they have to either be overwritten or raise an error if they exist (which we are saying is the default).
The two files which list data can either be overwritten with a new value, getting rid of all the old data, add the new value to it, overwriting a value that is already in it, or raise an error.
I did however just think of how error could possibly work for these files instead which may make more sense. We could get `write_mode = 'error'` to raise an error if the value is already in the file. That way it behaves pretty much the same way as `error` for the other files.

Reply:
sorry last clarification: what do you do when a file exists and `write_mode='append'`. Do you overwrite, raise an error, or skip writing the file?

Reply:
to quote the docstring
```
        - 'append': Any file that already exists will be overridden with new
            data, except `participants.tsv` and `scans.tsv` which will have the
            new data appended.
```

Reply:
So, let's say I already have `sub-01_ses-01_run-01_meg.fif` in my folder and the corresponding entry in `scans.tsv` and I run the script again with `'append`. In this case, do we overwrite the old entry in `scans.tsv` or append to it? Maybe that should also be clarified in the docstring?

Reply:
oh yeah, the old entry in the `scans.tsv` file will be overwritten.
cf. (in `_scans_tsv`):
```
        If write_mode == 'append' the existing file will have the
        new data added to it.
        If a duplicate entry is found it will be replaced with the new values
        provided.
```
I will add the last line of this to the docstring for `raw_to_bids` also.

Reply:
Okay I slept over it. I think we should raise an error in append mode if file already exists. And then we can make it the default. What do you think?

Reply:
> This is exactly what we shouldn't do. Otherwise what is the point of the append mode?

So, to me it seems that `append` should mean that you do not touch existing information. It is the least surprising API. I'm thinking of analogous mode in [open](https://docs.python.org/2/library/functions.html#open) for writing to files. (The file pointer starts at the end of the file). 

From what I understood, the benefit of having this mode was that you could write to partially written BIDS directory? If that is the case, why not simply ignore writing files that already exist? Instead of overwriting ... that way you don't have to deal with an error in the `tsv` file and you also won't need the `df.remove_duplicates` line (we should be as less dependent on `pandas` as possible).

Reply:
> If that is the case, why not simply ignore writing files that already exist?

Because then how will we ever add extra data to them?
The point is that we have the participants and scans.tsv files that contain information from multiple runs of MNE-BIDS.

Also, we could get rid of the drop duplicates line, but we'd need to replace it with a check to see if the data is already in the `df` and if so not write it. But I like the idea of being able to take the most recent piece of data.
So for example if you ran the script but forgot to set the birthday of the participant their age will be n/a. If you want to fix this and there are other exported data sets in this folder you don't want to run it with `overwrite` or `error`, but as it is if you run it with `append` (after adding the correct birthday), the data will all be correct.

Reply:
> The point is that we have the participants and scans.tsv files that contain information from multiple runs of MNE-BIDS.
> The point is that we have the participants and scans.tsv files that contain information from multiple runs of MNE-BIDS.

okay, but I would not call adding a line to these files as "overwrite". I would still call it "append". An overwrite for me would be if you completely rewrite the file.

> So for example if you ran the script but forgot to set the birthday of the participant their age will be n/a. If you want to fix this and there are other exported data sets in this folder you don't want to run it with overwrite or error, but as it is if you run it with append (after adding the correct birthday), the data will all be correct.

hmm ... in this case, I would say you need to use a text editor. Or -- once we have the functionality to read in BIDS folder to a `raw`, you could read it in, modify it in the `raw.info` and write it back.

Reply:
Oh wait, you're saying that `overwrite` would also overwrite `scans.tsv` ... so we can't write it back. Hmm ...

Reply:
yes. For files which can contain data from multiple runs `overwrite`, `append` and `error` all do distinctly different things. I feel what I wrote in the docstring should make it clear, or does it still not make sense?

```
    write_mode : str, one of ('append', 'overwrite', 'error')
        How the file should be handled if is exists already.
        Defaults to 'error'.
        If write_mode == 'append' the existing file will have the
        new data added to it.
        If a duplicate entry is found it will be replaced with the new values
        provided.
        If write_mode == 'overwrite' the existing file will be
        removed and replaced with the new data.
        If write_mode == 'error' an `OSError` will be raised if the data
        already exists in the file.
```

> hmm ... in this case, I would say you need to use a text editor. Or -- once we have the functionality to read in BIDS folder to a raw, you could read it in, modify it in the raw.info and write it back.

sure... or we could just use the `append` write mode as it is and it will do that for us.

Reply:
@monkeyman192 I think the docstring is useful once we have decided on the design. To me, it seems a mistake to overwrite when the mode is `append` even if the docstring says otherwise. We cannot expect people to read docstrings carefully. The API should be self-explanatory as much as possible. I have some new thoughts on the API for `overwrite`. Let me post it in a separate comment so we can discuss.

Reply:
see https://github.com/mne-tools/mne-bids/pull/99#issuecomment-430734213